src/HypreLinearSystem.C

```diff
@@ -140,83 +126,543 @@ HypreLinearSystem::beginLinearSystemConstruction()
+    std::vector<HypreIntType> hids(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
+    }
+
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+
+	const unsigned numNodes = (unsigned) b.num_nodes(k);
+	
+	if (numNodes) {
+	  stk::mesh::Entity const * nodes = b.begin_nodes(k);
+	  
+	  /* save the hypre ids */
+	  for (unsigned i=0; i<numNodes; ++i) hids[i] = get_entity_hypre_id(nodes[i]);
+	  
+	  /* fill owned/shared 1 Dof */
+	  fill_owned_shared_data_structures_1DoF(numNodes, hids);
+	}
+      }
+    }
+  } else {
+    std::vector<HypreIntType> hids(0);
+    std::vector<HypreIntType> columns(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
+      columns.resize(numNodes*numDof_);
+    }
+    
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+	
+	const unsigned numNodes = (unsigned) b.num_nodes(k);
+	
+	if (numNodes) {
+	  stk::mesh::Entity const * nodes = b.begin_nodes(k);
+	  
+          /* save the hids and columns */
+	  fill_hids_columns(numNodes, nodes, hids, columns);
+
+	  /* fill owned/shared for more than 1 Dof */
+	  fill_owned_shared_data_structures(numNodes, hids, columns);
+	}
+      }
+    }
+  }
+
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  gettimeofday(&_stop, NULL);
+  double msec = (double)(_stop.tv_usec - _start.tv_usec) / 1.e3 + 1.e3*((double)(_stop.tv_sec - _start.tv_sec));
+  printf("rank_=%d EqnName=%s : %s %s %d : dt=%1.5lf\n",rank_,name_.c_str(),__FILE__,__FUNCTION__,__LINE__,msec);
+#endif
 }
 
+
 void
-HypreLinearSystem::buildEdgeToNodeGraph(
-  const stk::mesh::PartVector&)
+HypreLinearSystem::buildEdgeToNodeGraph(const stk::mesh::PartVector& parts)
 {
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  /* record the start time */
+  struct timeval _start, _stop;
+  gettimeofday(&_start, NULL);
+#endif
+
   beginLinearSystemConstruction();
+  stk::mesh::MetaData & metaData = realm_.meta_data();
+  const stk::mesh::Selector s_owned = metaData.locally_owned_part()
+                                      & stk::mesh::selectUnion(parts)
+                                      & !(realm_.get_inactive_selector());
+  stk::mesh::BucketVector const& buckets = realm_.get_buckets(stk::topology::EDGE_RANK, s_owned);
+
+  if (numDof_==1) {
+    std::vector<HypreIntType> hids(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
+    }
+
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+
+	const unsigned numNodes = (unsigned) b.num_nodes(k);
```

Comment:
This numNodes can be obtained outside the loop (between lines 351 and 352). 'numNodes = b.topology().num_nodes();'. All entities within a bucket have the same topology, thus the same numNodes.

Reply:
Does this apply buildFaceToNodeGraph and buildElemToNodeGraph also?

Reply:
Yes the general rule is that within a bucket entities (faces, elements) all have the same number of nodes due to all having the same topology. But when an outer loop goes over multiple buckets, things can change from one bucket to the next. (Unless the selector that selected the buckets ensures that only homogeneous things are selected.)

Reply:
Ok. I've fixed buildEdgeToNodeGraph, buildFaceToNodeGraph and buildElemToNodeGraph to be:
  if (numDof_==1) {
    std::vector<HypreIntType> hids(0);

    for(size_t ib=0; ib<buckets.size(); ++ib) {
      const stk::mesh::Bucket & b = *buckets[ib];

      auto numNodes = b.topology().num_nodes();
      hids.resize(numNodes);

      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {

        stk::mesh::Entity const * nodes = b.begin_nodes(k);

        /* save the hypre ids */
        for (unsigned i=0; i<numNodes; ++i) hids[i] = get_entity_hypre_id(nodes[i]);

        /* fill owned/shared 1 Dof */
        fill_owned_shared_data_structures_1DoF(numNodes, hids);
      }
    }
  } else {
    std::vector<HypreIntType> hids(0);
    std::vector<HypreIntType> columns(0);

    for(size_t ib=0; ib<buckets.size(); ++ib) {
      const stk::mesh::Bucket & b = *buckets[ib];

      auto numNodes = b.topology().num_nodes();
      hids.resize(numNodes);
      columns.resize(numNodes*numDof_);

      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {

        stk::mesh::Entity const * nodes = b.begin_nodes(k);

        /* save the hids and columns */
        fill_hids_columns(numNodes, nodes, hids, columns);

        /* fill owned/shared for more than 1 Dof */
        fill_owned_shared_data_structures(numNodes, hids, columns);
      }
    }
  }

Reply:
Hopefully this is what you had in mind.

Reply:
Looks good.